tooling/src/tasks/distribute.ts

```diff
@@ -1,41 +1,108 @@
+import assert from "assert";
 import { task } from "hardhat/config";
 import { HardhatRuntimeEnvironment } from "hardhat/types";
 
-import { getProviders } from "../config";
+import { queryAmountToClaim } from "../queries/queryVestingPool";
+import calculateFundingBreakdown from "../fns/calculateFundingBreakdown";
+
+import {
+  checkpointCount,
+  checkpointExists,
+  loadSchedule,
+} from "../persistence";
+import { Schedule } from "../types";
+import { addresses, getProviders, VESTING_ID } from "../config";
 
 task(
-  "distribute",
-  "Creates a new checkpoint, calculates claim amounts, and updates distro setup enabling new claimers on both networks",
+  "distribute:prepare",
+  "Creates a new checkpoint, and persists it to the repo",
 ).setAction(async (_, hre: HardhatRuntimeEnvironment) => {
-  const providers = getProviders(hre);
-
   const log = (text: string) => console.info(`Task distribute -> ${text}`);
 
-  const {
-    isTokenReady,
-    isDistroReady,
-    isDelegateReady,
-    distroAddressMainnet,
-    distroAddressGnosis,
-  } = await hre.run("status", { silent: true });
-
-  if (!isTokenReady || !isDistroReady || isDelegateReady) {
-    log(
-      "Setup not ready for Distribution. Run status for more info. Skipping...",
-    );
+  const { isReady } = await hre.run("status", { silent: true });
+  if (!isReady) {
+    log("Setup not ready for Distribution. Skipping...");
     return;
   }
 
   await hre.run("harvest");
 
   await hre.run("schedule:validate", { frozen: true });
 
-  const { merkleRootMainnet, merkleRootGnosis } = await hre.run("checkpoint");
+  await hre.run("checkpoint", { persist: true });
+});
+
+task(
+  "distribute:apply",
+  "Calculates claim amounts, and posts the transactions that will eventually updates distro setup and enable new claimers",
+).setAction(async (_, hre: HardhatRuntimeEnvironment) => {
+  const log = (text: string) => console.info(`Task distribute -> ${text}`);
+
+  const { isReady, distroAddressMainnet, distroAddressGnosis } = await hre.run(
+    "status",
+    { silent: true },
+  );
+
+  if (!isReady) {
+    log("Setup not ready for Distribution. Skipping...");
+    return;
+  }
+
+  const schedule = loadSchedule();
+  const preCount = checkpointCount();
+  const { merkleRootMainnet, merkleRootGnosis } = await hre.run("checkpoint", {
+    persist: false,
+  });
+  const postCount = checkpointCount();
+  // its important to ensure that checkpoint:apply is running
+  // on artifacts previously calculated and committed
+  assert(preCount === postCount, "Checkpoint task persist=false is writing?");
```

Comment:
this assert message is hard to parse, I'm not sure what it's saying

Reply:
Right. So it's about ensuring that the checkpoint subtask calling during the apply pass is calculating the checkpoint but not persisting it (ensure that the persist:false is really taking effect). So it should have been priorly persisted. Its really coming from an abundance of caution, and preventing some scenario where a code refactoring changes this.

More context:
- distribute was split in prepare and apply
- prepare is to calculate and persist the checkpoint
- apply is to actually propose to safes
- if apply fails, and prepare doesn't, no biggie we have an extra checkpoint in the codebase no harm is done
- the reserve, i.e., if we post something to a safe that was not persisted, it's kind of a messy situation that will require manual resolution and will result in a broken claim frontend